src/types.md

```diff
@@ -0,0 +1,183 @@
+    let mut a = ObjectA::new(&mut shared_thingy);
+    a.do_something();
+}
+```
+
+Do this:
+
+```rust
+struct ImportantSharedObject;
+
+struct ObjectA;
+
+impl ObjectA {
+    fn new() -> Self {
+        Self
+    }
+    fn do_something(&mut self, important_shared_object: &mut ImportantSharedObject) {
+        // act on important_shared_object
+    }
+}
+
+fn main() {
+    let mut shared_thingy = ImportantSharedObject;
+    let mut a = ObjectA::new();
+    a.do_something(&mut shared_thingy);
+}
+```
+
+(happily this also gets rid of named lifetime parameters.)
+
+If you have a hundred such shared objects, you probably don't want a
+hundred function parameters. So it's usual to bundle them up into
+a context structure which can be passed into each function call:
+
+```rust
+struct ImportantSharedObject;
+struct AnotherImportantObject;
+
+struct Ctx<'a> {
+    important_shared_object: &'a mut ImportantSharedObject,
+    another_important_object: &'a mut AnotherImportantObject,
+}
+
+struct ObjectA;
+
+impl ObjectA {
+    fn new() -> Self {
+        Self
+    }
+    fn do_something(&mut self, ctx: &mut Ctx) {
+        // act on ctx.important_shared_object and ctx.another_important_thing
+    }
+}
+
+fn main() {
+    let mut shared_thingy = ImportantSharedObject;
+    let mut another_thingy = AnotherImportantObject;
+    let mut ctx = Ctx {
+        important_shared_object: &mut shared_thingy,
+        another_important_object: &mut another_thingy,
+    };
+    let mut a = ObjectA::new();
+    a.do_something(&mut ctx);
+}
+```
+
+```mermaid
+flowchart LR
+    subgraph Shared functionality
+    important[Important Shared Object]
+    end
+    subgraph Context
+    refa[Mutable Reference]-->important
+    end
+    subgraph ObjectA
+    objectA[Object A]
+    methodA[Method]
+    methodA-. Acts on shared object.->important
+    end
+    subgraph ObjectB
+    objectB[Object B]
+    methodB[Method]
+    methodB-. Acts on shared object.->important
+    end
+    main --> objectA
+    main --> objectB
+    main --> Context
+    main-. Passes context.-> methodA
+    main-. Passes context.-> methodB
+```
+
+Even simpler: just put all the data directly into `Ctx`. But the key point
+is that this context object is passed around into just about all function calls
+rather than being stored anywhere, thus negating any borrowing/lifetime concerns.
+
+This pattern can be seen in, for example, [bindgen](https://github.com/rust-lang/rust-bindgen/blob/271eeb0782d34942267ceabcf5f1cf118f0f5842/src/ir/context.rs#L308).
+
+> Split out borrowing concerns from the object concerns. - MG
+
+(To generalize this idea, try to avoid storing references to anything that might
+need to be changed. Instead take those things as parameters. For instance
+`petgraph` [takes the entire graph as context to a `Walker` object](https://docs.rs/petgraph/0.6.0/petgraph/visit/trait.Walker.html), such that the graph can be changed whilst you're walking it.)
```

Comment:
"whilst" is British English; use "while" instead.

(I'm assuming you want to use American English? For example, you've got "behavior" above.)